lib/src/wallet/data.rs

```diff
@@ -189,6 +196,58 @@ impl WitnessCache {
     //     return hex::encode(buf);
     // }
 }
+pub(crate) trait FromCommitment {
+    fn from_commitment(from: &[u8; 32]) -> Self;
+}
+
+pub(crate) trait ToBytes<const N: usize> {
+    fn to_bytes(&self) -> [u8; N];
+}
+
+impl ToBytes<32> for SaplingNullifier {
+    fn to_bytes(&self) -> [u8; 32] {
+        self.0
+    }
+}
+
+impl ToBytes<32> for OrchardNullifier {
+    fn to_bytes(&self) -> [u8; 32] {
+        OrchardNullifier::to_bytes(*self)
+    }
+}
+
+impl FromCommitment for SaplingNode {
+    fn from_commitment(from: &[u8; 32]) -> Self {
+        Self::new(*from)
+    }
+}
+impl FromCommitment for MerkleHashOrchard {
+    fn from_commitment(from: &[u8; 32]) -> Self {
+        Self::from_bytes(from).unwrap()
+    }
+}
+
+pub(crate) trait NoteData {
+    type Fvk: Clone;
+    type Diversifier;
```

Comment:
`Diversifier` and `Note` are associated types of the `NoteData` trait (which is implemented on `OrchardNoteData` and `SaplingNoteData`).

But `struct SaplingNoteData` has fields `pub note` and `pub diversifier`.
`struct OrchardNoteData` has fields `diversifier` and `note` as well.

This smells suspicious to me..   can we not DRY these fields somehow.

Also, given the adjacent declarations and implementations...   should we simplify the `struct` names to `OrchardNote` and `SaplingNote`?

Reply:
I gave myself a non-blocking issue to follow up on this:

https://github.com/zingolabs/zingolib/issues/57